File: haskell/bob/src/Bob.hs
module Bob (responseFor) where
import Data.Char (isUpper, isLetter, isSpace)
import Data.String.Utils 

isQuestion :: String -> Bool
isQuestion s = '?' == (last . strip) s

isYelled :: String -> Bool
isYelled s = let normS = filter (isLetter) s in
  length normS == (length $ filter (isUpper) normS) && length normS > 0

responseFor :: String -> String
responseFor xs
  | null $ filter (not . isSpace) xs = "Fine. Be that way!"
  | isQuestion xs && isYelled xs = "Calm down, I know what I'm doing!"
  | isQuestion xs = "Sure."
  | isYelled xs = "Whoa, chill out!"
  | otherwise = "Whatever."

File: scala/bob/src/main/scala/Bob.scala
object Bob {
  def isQuestion(statement: String): Boolean = statement.takeRight(1) == "?"
  def isShouted(statement: String): Boolean = {
    // x.isLetter => x.isUpper AND statement contains a letter
    statement.foldRight(true) ((x: Char, y: Boolean) => (!x.isLetter || x.isUpper) && y) &&
    statement.foldRight(false) (_.isLetter || _)
  }
  def isNothing(statement: String): Boolean = (statement == "") || (statement.foldRight(true) (!_.isLetterOrDigit && _))

  def response(statement: String): String = {
    val s = statement.trim
    if (isQuestion(s) && isShouted(s)) "Calm down, I know what I'm doing!"    
    else if (isQuestion(s)) "Sure."
    else if (isNothing(s)) "Fine. Be that way!"
    else if (isShouted(s)) "Whoa, chill out!"
    else "Whatever."
  }
}
